src/contract.rs

```diff
@@ -242,11 +248,98 @@ pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryM
+        return Err(StdError::generic_err(format!(
+            "Must sign exactly 1 permit message, got: {:?}.",
+            permit.signed.msgs.len()
+        )));
+    }
+
+    if permit.signed.msgs[0].r#type != "query_permit" {
+        return Err(StdError::generic_err(format!(
+            "Type must be 'query_permit', got: {:?}.",
+            permit.signed.msgs[0].r#type
+        )));
+    }
+
+    let permit_content = &permit.signed.msgs[0].value;
+
+    // Validate permit_name
+    let _permit_name = &permit_content.permit_name;
+    // TODO fail if permit_name is revoked
+
+    // Validate permit message
+    let token_address = ReadonlyConfig::from_storage(&deps.storage)
+        .constants()?
+        .contract_address;
+
+    if !permit_content.allowed_tokens.contains(&token_address) {
+        return Err(StdError::generic_err(format!(
+            "Permit doesn't apply to token {:?}, allowed tokens: {:?}",
+            token_address, permit_content.allowed_tokens
+        )));
+    }
+
+    // Derive account from pubkey
+    let pubkey = permit.signature.pub_key.value;
+    let account = deps.api.human_address(&pubkey_to_account(&pubkey))?;
+
+    // Validate signature, reference: https://github.com/enigmampc/SecretNetwork/blob/f591ed0cb3af28608df3bf19d6cfb733cca48100/cosmwasm/packages/wasmi-runtime/src/crypto/secp256k1.rs#L49-L82
+    let signed_bytes = &to_binary(&permit.signed)?.0;
+    let signed_bytes_hash = Sha256::digest(signed_bytes);
+    let secp256k1_msg =
+        secp256k1::Message::from_slice(signed_bytes_hash.as_slice()).map_err(|err| {
+            StdError::generic_err(format!(
+                "Failed to create a secp256k1 message from signed_bytes: {:?}",
+                err
+            ))
+        })?;
+
+    let secp256k1_verifier = Secp256k1::verification_only();
+
+    let secp256k1_signature = secp256k1::Signature::from_compact(&permit.signature.signature.0)
+        .map_err(|err| StdError::generic_err(format!("Malformed signature: {:?}", err)))?;
+    let secp256k1_pubkey = secp256k1::PublicKey::from_slice(pubkey.0.as_slice())
+        .map_err(|err| StdError::generic_err(format!("Malformed pubkey: {:?}", err)))?;
+
+    secp256k1_verifier
+        .verify(&secp256k1_msg, &secp256k1_signature, &secp256k1_pubkey)
+        .map_err(|err| {
+            StdError::generic_err(format!(
+                "Failed to verify signatures for the given permit: {:?}",
+                err
+            ))
+        })?;
+
+    match query {
+        QueryWithPermit::Balance {} => query_balance(&deps, &account),
+        QueryWithPermit::TransferHistory { page, page_size } => {
+            query_transfers(&deps, &account, page.unwrap_or(0), page_size)
+        }
+        QueryWithPermit::TransactionHistory { page, page_size } => {
+            query_transactions(&deps, &account, page.unwrap_or(0), page_size)
+        }
+        QueryWithPermit::Allowance { spender } => query_allowance(deps, account, spender),
```

Comment:
in allowance it's important to see all allowances given to the querier, as well as all allowances the querier has granted,  which is why there's the for loop in `viewing_keys_queries`. This only allows seeing the allowances that the querier has given.

Reply:
Please confirm that c3071e908616d8fb168a86daf67f925d2635e753 fixed the issue.

Reply:
Yup, We'll have to test it but i believe it works!